### ISSUE TITLE
Claude Code 연동: capture 의 11초는 기동 비용이고, MCP/훅의 namespace 기본값이 어긋난다

### ISSUE THREAD

**Issue by jinmang2** [jinmang2/agentic-memory#2]
Claude Code 에 실제로 물려보다 두 가지가 걸렸다. 하나는 성능이고 하나는 정합성인데, **붙이기 전에 정해야 하는 쪽은 두 번째**다.

측정 환경: WSL2 · 4 cores · RAM 7.76 GB · RTX 2060 6 GB · `--profile lite` · `SqliteVecStore` · `intfloat/multilingual-e5-small` · **LLM 엔드포인트 없음** (`capabilities.json` 의 세 후보 전부 `alive: false`). 임시 `--data-dir` 로만 돌렸다.

---

## 1. capture 의 11초는 일이 아니라 기동 비용이다

`hooks/capture.py` 의 docstring 이 적어둔 10.8 s 를 재현했고(11.30 s / 10.93 s), **같은 쓰기를 떠 있는 서버에 시키면 41–62 ms** 다.

| 경로 | 시간 |
|---|---|
| `python -m agmem.hooks.capture` (프롬프트마다 새 프로세스) | **11,300 ms** |
| 상주 서버에 `add_memory` (HTTP, 웜) | **41–62 ms** |

내역:

```
import torch                   1.30 s
import sentence_transformers   3.84 s
모델 구성                      ~3.7 s   (docstring 측정치)
인터프리터 + agmem 초기화       ~2 s
─────────────────────────────────────
실제 임베딩 + 쓰기              ~0.05 s
```

`recall` 이 0.16 s 로 싼 것도 같은 사실의 뒷면이다 — 임베더를 안 쓰고 문서 스토어만 읽는다.

```
recall  0.19 / 0.16 / 0.16 s   (592 B 주입, SessionStart timeout 10 s 의 2%)
```

**`--transport http` 가 이미 그 "long-lived process" 다.** docstring 이 정직한 해법으로 지목한 것이 이미 절반 만들어져 있다:

```
서버 기동      45 s (한 번)
initialize     33 ms
add_memory     1677 → 62 → 41 → 50 ms   (첫 호출은 임베더 지연 로드)
search_memory  1101 → 31 ms
memory_stats    5 ms
```

MCP stdio 도 세션마다 초기화를 낸다 — 최초 28.6 s, 이후 8.9–10.0 s.

그래서 상주 서비스 하나에 소비자 셋이 붙는 그림이 가능하다:

```
agmem (HTTP, 상주)
  ├── Claude Code MCP        세션당 10 s 초기화 제거
  ├── capture 훅 = POST      11,300 ms → ~50 ms
  └── recall  훅 = GET          160 ms → ~30 ms
```

**공짜는 아니다.** 지금 설계의 미덕은 상주 프로세스가 *없다*는 것이고, 데몬을 두면 없던 실패 모드가 생긴다 — 데몬이 죽어 있으면 capture 가 조용히 턴을 잃는다. 훅이 자기 완결적일 때는 없는 문제다. 그러니 이건 "빨라진다"가 아니라 **11 초를 감수할 것인가 vs 라이프사이클을 감수할 것인가**의 교환이고, 후자를 고른다면 헬스체크와 "서버가 없으면 어떻게 할 것인가"(조용히 버림 / 로컬 폴백 / 나중에 배치 임베딩)를 함께 정해야 한다.

## 2. MCP 와 훅의 namespace 기본값이 어긋난다 ← 데이터를 쓰기 전에 정해야 하는 쪽

```
mcp/server.py:163   ap.add_argument("--namespace", default="main")
hooks/__init__.py:37  DEFAULT_NAMESPACE = "claude-code"
```

둘을 그대로 켜면 **MCP 는 `main` 에 쓰고 recall 은 `claude-code` 를 읽는다.** 서로 못 본다. 에러 없이, 그냥 아무것도 회수되지 않는다 — 조용한 실패라 알아채는 데 시간이 걸린다.

더 구조적인 쪽: **`namespace` 는 서버 기동 플래그이고 도구 인자가 아니다.**

```
add_memory(content: str, role: str = "user", timestamp: str | None = None)
search_memory(...)            # 둘 다 namespace 를 받지 않는다
```

그래서 **데몬 하나 = 메모리 하나**다. 프로젝트별로 기억을 나누려면 프로젝트마다 데몬을 띄우거나 전부 한 통에 섞어야 한다. 섞은 뒤에 나누는 건 훨씬 비싸므로, 쓰기를 시작하기 전에 결정하는 게 맞다.

생각해볼 갈래:
- 도구가 `namespace` 를 선택 인자로 받게 한다 (데몬 하나로 다중 프로젝트)
- 최소한 두 기본값을 일치시키고, 불일치를 기동 시 경고한다
- 아니면 "한 통"을 명시적 설계로 선언하고 문서에 못박는다

## 재현

```bash
# 훅 두 개
AGMEM_DATA_DIR=/tmp/agmem-probe AGMEM_NAMESPACE=probe \
  printf '{"session_id":"x","cwd":"'$PWD'","prompt":"t","hook_event_name":"UserPromptSubmit"}' \
  | time .venv/bin/python -m agmem.hooks.capture

# HTTP 상주 서버
.venv/bin/agmem-mcp --profile lite --transport http --port 8931 \
  --data-dir /tmp/agmem-probe --namespace probe
# POST http://127.0.0.1:8931/mcp  (streamable-http, mcp-session-id 헤더 유지)
```

## 남는 것

측정값 전부 **LLM 없는 경로**다. `[llm.extract]` 를 붙이면 organizer 가 쓰기마다 LLM 을 호출하므로(4 에피소드에 `evolution_ops: 4` 확인) 쓰기 지연은 완전히 달라진다. LLM 을 붙일 계획이면 그 상태로 다시 재야 한다.

**Comment by jinmang2** [comment]
## §2 (namespace 불일치) — 39356f1 에서 해결. §1 (기동 비용) — 미해결, 열어 둠.

**§2 조치**
- `agmem.env` 신설: namespace / data_dir / config 해석과 기본값을 한 곳에. facade·hooks·MCP 서버가 전부 여기서 import. 기본 namespace는 양쪽 다 `main`.
- MCP 서버가 `AGMEM_NAMESPACE` / `AGMEM_DATA_DIR` / `AGMEM_CONFIG` 를 읽는다 (이전엔 환경변수를 전혀 안 읽었음). 우선순위: CLI > env > `agmem.toml` > 기본값. 훅과 같은 변수이므로 **한 번 export 하면 두 층이 같은 스토어를 연다.**
- **모든 도구가 선택 인자 `namespace`** 를 받는다. 서버 하나가 namespace 를 지연 개방해 유지하고 임베더는 공유 — 데몬 하나 = 메모리 한 통이던 제약 해소. 값은 경로 조각으로 검증 (`../x` 거부).
- 훅에 `AGMEM_CONFIG` 이음매 추가 (서버 `--config` 와 같은 파일, organizer 는 여전히 빈 목록).

**이슈가 못 본 것 2건도 같이**
- `scripts/smoke_product_stack.py` 가 양쪽에 같은 namespace 를 명시해서 이 불일치를 구조적으로 못 봤음 → 이제 어느 쪽에도 안 알려주고 디렉터리 == 서버 보고 namespace 를 판정에 포함. 실측(조용한 머신): capture 12.9 s, recall 0.22 s, handshake 9.7 s, 양쪽 `main`.
- `tests/test_hooks.py` 가 실행마다 471 MB 모델을 내려받고 있었음 (HOME 이 tmp 라 HF 캐시가 안 보임 + FakeEmbedder 주입 이음매 부재). 이날 120 s 타임아웃으로 2 ERROR → `AGMEM_CONFIG` 로 밀폐, 32 tests 5.7 s.

문서: docs/05 §2.2 (변수 표), §2.3 (등록 예시를 env 방식으로), §2.4. 스위트 886 passed.

**§1 은 그대로다.** capture 가 프로세스당 임베더를 올리는 구조는 바뀌지 않았고, 이 머신에서 오늘 재측정하니 웜 ~13–19 s, 콜드 캐시 직후 56 s 로 이슈의 11 s 보다 머신 상태에 더 민감했다 — 상주 프로세스 논거를 오히려 강화한다. 데몬 전환은 라이프사이클·헬스체크·부재 시 동작을 먼저 정해야 하는 설계 결정이라 별도로 진행한다.

**Comment by jinmang2** [comment]
## §1 (기동 비용) — d63a7fd 에서 해결. 이슈의 두 절이 모두 닫혀 클로즈한다.

이슈가 지목한 대로 `agmem-mcp --transport http` 가 상주 프로세스가 됐다. 훅용 평문 라우트 `GET /health` / `POST /hooks/capture` / `POST /hooks/recall` 을 열고, 훅은 stdlib `urllib` 만으로 부른다 (훅 프로세스에 MCP SDK 도 torch 도 들어오지 않는다).

**라이프사이클**: SessionStart 훅이 `/health` 실패 시 detach 로 스폰, `--idle-timeout`(기본 1800 s) 후 자체 종료. 두 호스트(Claude Code·Codex)가 같은 환경변수로 같은 데몬을 공유.

**부재 시 동작** — 이슈가 "가짜 빠름" 이라 경계한 지점: 훅은 모델을 올리지 않는다. capture 는 에피소드를 doc store 에만 쓰고(벡터 없음) 데몬을 요청한 뒤 exit 0; 데몬이 기동 시와 매분 "doc store 에 있고 vector store 에 없는" 에피소드를 백필한다. 벡터가 영구히 비는 경우는 없고, `/health` 의 `pending_embed` 가 지연을 드러낸다. 스모크가 잡은 구멍 하나: 데몬이 한 번도 안 뜬 머신에서 Claude Code 가 stdio 로 띄우는 MCP 서버는 그 에피소드를 영영 못 봤을 것 → **모든 전송에서 서버 기동 시 백필 1회**.

**신규 훅 `recall_prompt`** (UserPromptSubmit): 프롬프트를 질의로 top-5 를 `additionalContext` 로 주입. 이 리포의 LME 실측(읽기 경로 +21.2pp, docs/21)과 제품군 조사(docs/research/product-memory-landscape.md)가 근거.

**실측** (실제 임베더, 전체 스위트가 병행 중이던 머신):

| 경로 | 이슈 시점 | 지금 |
|---|---|---|
| capture, 데몬 없음 | 11.3 s (이날 재측정 12.9–19 s, 콜드 56 s) | **0.8–1.4 s** |
| capture, 데몬 웜 | 41–62 ms (서버 내부) | **0.46–0.58 s** (훅 프로세스 기동 포함) |
| recall_prompt, 웜 | — | 0.8–1.6 s, found |
| 데몬 기동 | 45 s | 25–46 s (부하), 조용하면 ~10 s |

웜 capture 의 남은 0.4 s 는 임베더가 아니라 훅 프로세스의 인터프리터 기동 + `import agmem`(패키지 `__init__` 이 facade 를 끌어옴, 0.59 s) 이다. 패키지 지연 import 로 줄일 수 있는 별건이라 여기서 다루지 않았다.

테스트 `tests/test_daemon.py` 7 건(실제 데몬 subprocess + 실제 훅 subprocess, FakeEmbedder 밀폐), 스위트 877/2, `scripts/smoke_product_stack.py --daemon` 양 경로 ok, 잔류 데몬 없음. 문서 docs/05 §2.3.1 신설·§2.4.
